pilot/pkg/model/sidecar.go

```diff
@@ -447,10 +448,11 @@ func convertIstioListenerToWrapper(ps *PushContext, configNamespace string,
 			out.listenerHosts[parts[0]] = make([]host.Name, 0)
 		}
 		out.listenerHosts[parts[0]] = append(out.listenerHosts[parts[0]], host.Name(parts[1]))
+		listenerHosts = append(listenerHosts, host.Name(parts[1]))
```

Comment:
By only taking into account the hostname and not namespace, we end up importing virtual services from namespaces we shouldn't. Take this test case for example:

```go
		{
			name: "sidecar filter",
			cfg: "" +
				vs(t, vsArgs{
					Namespace: "not-default",
					Match:     "*.example.com",
					Dest:      "arbitrary.example.com",
				}) +
				vs(t, vsArgs{
					Namespace: "default",
					Match:     "explicit.example.com",
					Dest:      "explicit.example.com",
				}) +
				sidecarScope(t, scArgs{
					Namespace: "not-default",
					Egress:    []string{"not-default/*.example.com", "not-default/not-default.example.org"},
				}),
			proxy:     proxy("not-default"),
			routeName: "80",
			expected: map[string][]string{
				// even though there is an *.example.com, since we do not import it we should create a wildcard matcher
				"*.example.com": {"outbound|80||arbitrary.example.com"},
				"explicit.example.com": nil,
			},
		},
```

With this PR, we end up with a route for `explicit.example.com`. This should not happen since its from `default` namespace which we do not import

Reply:
I added all the tests in https://github.com/istio/istio/pull/37813. can you pull those into this PR?

Reply:
>By only taking into account the hostname and not namespace, we end up importing virtual services from namespaces we shouldn't.

Listener hosts namespace has nothing todo with vritualservice namespace, see the bug reported.

Reply:
>I don't think that's true. Sidecar scope DOES apply the VS.

Nope, Where is the doc saying that?

Reply:
"or a service specified using a ServiceEntry or VirtualService" 
https://istio.io/latest/docs/reference/config/networking/sidecar/#IstioEgressListener

beside, regardless if doc this is how it has behaved since Istio 1.1

Reply:
It does not say the namespace should equal to vs namespace

Because we can define a vs in ns A, but it can point to a service in namespace B

Reply:
This is how it behaved in Istio 1.1 and beyond. If you want to change the behavior of the established API maybe it would be preferred to start with a design doc explaining the use cases, a migration path, etc. Without that I would recommend we fix the regression and go back to the original behavior as documented.

Reply:
I donot think this is changing semantic, it is a bug since beginning.

Reply:
The bugs is hidden because , sidecar egress hosts is not widely used in combine with vs

Reply:
I don't understand why you think it's a bug? 

The ONLY place explicit.example.com.is defined is in ns "default"
 the proxy ONLY imports from namespace "not-default"

so why should that proxy see explicit.example.com?

Reply:
I believe the behavior prior to these changes (eg in 1.11) was correct, and changes to it should to through a design doc

Reply:
explicit.example.com is the vs host right, but how about the service is in `node-default` namespace.

Reply:
cc @costinm may participate the original discussion

Reply:
There is no service in not-default called "explicit.example.com". That is only defined in "default" VirtualService

Reply:
Yes, but waht if the svc is in `not-default`

Reply:
This is the scenario the issue reported

Reply:
In the users case they are importing everything. In the test above we are trying to restrict what is imported

So in user case, we restricted too much. But now we do not restrict enough in this PR

Reply:
Oh, i get to know the issue of this pr

Reply:
@howardjohn I have to clarify there are 2 rounds of  vs selecting logic. 

1. when we build sidecarscope  --> selected `vs sets 1`
2.  when we build route, like for 8080, only for service with 8080. This time we  select vses from the  `vs sets 1`

Reply:
In your tests, it is route 80, which is very different. You can see the comments I added here 

https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/httproute.go#L274

Reply:
https://github.com/istio/istio/pull/33764#issuecomment-873505981